crates/kitsune_p2p/types/src/tx2/tx2_utils/notify_all.rs

```diff
@@ -50,7 +84,29 @@ impl Default for NotifyAll {
 impl NotifyAll {
     /// Construct a new NotifyAll instance.
     pub fn new() -> Self {
-        Self(Arc::new(Share::new(Vec::new())))
+        let inner = Inner {
+            wakers: Vec::new(),
+            cbs: Vec::new(),
+        };
+        let wrap = Arc::new(Share::new(inner));
+        let notify_on_drop = Arc::new(NotifyOnDrop(wrap.clone()));
+        Self(wrap, notify_on_drop)
+    }
+
+    /// Register a sync cb to be invoked on notify
+    /// (will be invoked immediately if did_notify())
+    pub fn wait_cb<F>(&self, sync_cb: F)
+    where
+        F: FnOnce() + 'static + Send,
+    {
+        let mut maybe_sync_cb: Option<NotifySyncCb> = Some(Box::new(sync_cb));
+        let _ = self.0.share_mut(|i, _| {
+            i.cbs.push(maybe_sync_cb.take().unwrap());
```

Comment:
Why is `maybe_sync_cb` an option? Does the closure in `share_mut` only ever get called at max once so the unwrap is safe?
Is there a chance `share_mut` won't call the closure and that's why there is an if let below? 
If this is the case maybe some comments would be worth it to help with understanding.

Reply:
Yeah, exactly - If we moved sync_cb into the share_mut closure, and we had already notified, share_mut would return an error and drop sync_cb so we wouldn't be able to call it.

Added some comments for clarity.